File: c-codes/sorting/mergesort/mergesort.cpp
/**
 * @Author: @IamRezaMousavi
 * @Date:   2021-12-16 17:18:37
 * @Last Modified by:   @IamRezaMousavi
 * @Last Modified time: 2021-12-17 01:26:37
 */
#include <iostream>

using namespace std;

void merge(int array[], int left, int mid, int right);
void mergeSort(int array[], int begin, int end);

int main(int argc, const char *argv[]) {
  int array[]   = {10, 8, 5, -2, 17, 15, 22, -4};
  int arraySize = sizeof(array) / sizeof(array[0]);

  for (size_t i = 0; i < arraySize; i++)
	cout << array[i] << " ";
  cout << endl;

  mergeSort(array, 0, arraySize); // array[begin:end]

  for (size_t i = 0; i < arraySize; i++)
	cout << array[i] << " ";

  return 0;
}

void merge(int array[], int left, int mid, int right) {
  int leftIndex = 0, rightIndex = 0, arrayIndex = left;
  int leftSize = mid - left, rightSize = right - mid;
  int leftArray[leftSize], rightArray[rightSize];

  for (size_t i = 0; i < leftSize; i++)
	leftArray[i] = array[left + i];
  for (size_t i = 0; i < rightSize; i++)
	rightArray[i] = array[mid + i];

  while (leftIndex < leftSize && rightIndex < rightSize) {
	if (leftArray[leftIndex] <= rightArray[rightIndex]) {
	  array[arrayIndex] = leftArray[leftIndex];
	  leftIndex++;
	} else {
	  array[arrayIndex] = rightArray[rightIndex];
	  rightIndex++;
	}
	arrayIndex++;
  }

  while (leftIndex < leftSize) {
	array[arrayIndex] = leftArray[leftIndex];
	leftIndex++;
	arrayIndex++;
  }
  while (rightIndex < rightSize) {
	array[arrayIndex] = rightArray[rightIndex];
	rightIndex++;
	arrayIndex++;
  }
}

void mergeSort(int array[], int begin, int end) {
  if (end - begin == 1) return;

  int mid = (end + begin) / 2;

  mergeSort(array, begin, mid); // left = array[begin:mid]
  mergeSort(array, mid, end);   // right = array[mid:end]
  merge(array, begin, mid, end);
}

File: sorting/mergesort/mergesort.py
# @Author: @IamRezaMousavi
# @Date:   2021-12-16 03:26:20
# @Last Modified by:   @IamRezaMousavi
# @Last Modified time: 2021-12-16 03:42:27


def merge(list1, list2):
    sizeList1 = len(list1)
    sizeList2 = len(list2)
    _list = []

    i = j = 0
    while i < sizeList1 and j < sizeList2:
        if list1[i] < list2[j]:
            _list.append(list1[i])
            i += 1
        else:
            _list.append(list2[j])
            j += 1
    _list += list1[i:]
    _list += list2[j:]
    return _list


def mergeSort(_list):
    if len(_list) == 1:
        return _list
    mid = len(_list) // 2
    right = mergeSort(_list[mid:])
    left = mergeSort(_list[:mid])
    return merge(right, left)


print(mergeSort([1, 0, 2, 4, 5, -2, -11, 7, 11]))
